### ISSUE TITLE
Memory Leak.

### ISSUE THREAD

**Issue by RoiEXLab** [triplea-game/triplea#6050]
Basically #4798 - Whenever you start a local game and close it, the underlying TripleAFrame will never get garbage collected.
Not a problem for bot games because those will launch a new JVM and kill it once that's over so the memory is freed by the OS.
Somewhere there's a reference to it; My best guess is that we call some swing helper at some point that adds a reference to this frame in a static list somewhere in swing code.
Just wasted another hour trying to get behind this, but no luck.

Interestingly enough it doesn't matter if you use the current swing implementation to launch the game, or use the experimental JavaFX client, the Frame keeps existing.

This completely unsasitfying, and currently my best approach to this is trial and error, which due to the complexity of the class structure isn't productive at all.
So it shouldn't be a surprise to you that I'm currently not willing to waste any more time on this stupid issue.

So, do you have any idea what could be done in order to fix this issue? Any ideas on how to narrow down the issue to a specific class? A specific method? @DanVanAtta @ssoloff @ron-murhammer

**Comment by DanVanAtta** [comment]
Not necessarily an easy problem. My 2 cents would be to focus on simplifications/refactoring. Eventually either that would fix the problem outright or at least it'll be much easier to reason about the code and far easier to spot.

Side-topic: It would be really great to fix the launching in a separate process. We want to fix that so we can list on steam. Secondarily, I've noticed debuggers do not always trigger on the new process that is launched, can make debugging very difficult sometimes.

**Comment by RoiEXLab** [comment]
Yeah, noticed the debugger thing as well, it's annoying

**Comment by Cernelius** [comment]
@RoiEXLab I can tell that if I'm mapmaking, changing the xml, and keeping closing and reopening the game, without closing and reopening TripleA, it happened several times to me that, after some open and close, I will hit "overead" "GC Limit" something (not sure), the game failing to load, obliging me closing that instance of TripleA. So, basically, something major of the previous game I've closed must be lingering there, for no reasons.

Is this the same issue at hand, or am I being off topic?

Also, would this be a problem for regular hosts, but not for bots, as I understand it (keeping increasing the RAM until it is too big and the game fails, if you keep end and restarting the game, without closing and reopening the host itself)?

**Comment by RoiEXLab** [comment]
> So, basically, something major of the previous game I've closed must be lingering there, for no reasons.

Almost everything seems to stay there, but we have no idea why.

> Also, would this be a problem for regular hosts, but not for bots [...]

Interesting consideration, I did not think about that.
However this is a UI related Problem (it seems), so without any UI this is not a problem. On the other hand, if you host a game in the lobby, this might be a problem, however once you close the window and click on the host button again you create a clean new process without uneccessary memory.

**Comment by RoiEXLab** [comment]
> Is this the same issue at hand, or am I being off topic?


In case it wasn't obvious enough: Yes, this is the exact issue this ticket is about

**Comment by Cernelius** [comment]
@RoiEXLab Ok so, if it can of help to find the root of the matter, another thing I've noticed, and that you can test, as I say, is that if you open TripleA and start a game, then close the game, without closing TripleA, then open the map (assuming you are not looking at a mapskin) and change some unit image files, so that the units will look different, then restart a new game, you will have the new current images on the board view, the battle window, the battlecalculator and the unit help, while, if you look at the territory tab, the old images (that you have deleted, so don't exist anymore in your HD) still displays there.

So, I think the "Territory" tab (only?) has this weird problem of keeping in memory images after you close the game. So, in theory, if a mapmaker would update a map unit images while a user is playing that map, and, then, the user update and restart without closing that same istance of TripleA, the "Territory tab" images would be the old ones while the rest are the new ones (possibly confusing).

Maybe this is one of the roots of stuff being kept in memory?

I reported this in the past, but I believe it was just closed, nothing coming out of it. This strange/inconsistent behaviour exists since I started mapmaking, that is now many years.

**Comment by RoiEXLab** [comment]
@Cernelius Not sure if this is related, but it it's definitely something worth investigating

**Comment by RoiEXLab** [comment]
I have to admit that this is a highly complex issue and therefore not something feasible to tackle.

I want to point out that this is an important issue nevertheless. If playing a big map like TWW, just restarting the game (because you forgot to set a map specific setting, or messed up in AI assignment) will silently consume about 1 GB on my machine in RAM.
Not a huge deal if you have 12GB or more on your windows machine, but on systems with 4 or 8GB that's quite a lot to throw away.

It's worth noting that this issue continues to be the case in the JavaFX client, so it definitely has something to do with the swing implementation of the main game (not the launching screen) storing a reference of itself somewhere.
I wonder if we could remove component after component until the issue goes away to identify which component is introducing the problem, but at this point I'm not in the mood to waste another couple of hours of my lifetime for this issue.

**Comment by DanVanAtta** [comment]
IF we could be more systematic, perhaps our efforts would combine. Want to point out spending time trying to fix a problem is not a waste. The battle screen pop-up error certainly took about 3 different long debugging sessions to final find it.

**Comment by RoiEXLab** [comment]
@DanVanAtta You are correct, but after spending way too many hours on this I simply don't _feel_ like it was worth it
